### DISCUSSION TITLE
Question: temporary struct lifetime

### DISCUSSION THREAD

**Discussion by kevinjwz** [dotnet/csharplang#3404] [Q&A]
I am writing a value-type list implicitly convertible to `Span<T>`, as a temporary solution for "params Span<T>". 
As a first try:
```csharp
[StructLayout(LayoutKind.Sequential)]
struct ValueList
{
	int length;
	int i0,i1,i2,i3;

	public static implicit operator Span<int>(in ValueList v) =>
		MemoryMarshal.CreateSpan(ref Unsafe.AsRef(v.i0), v.length);	

	public static ValueList List(int i0) => 
		new ValueList { length = 1, i0 = i0 };

	public static ValueList List(int i0, int i1) => 
		new ValueList { length = 2, i0 = i0, i1 = i1 };

	public static ValueList List(int i0, int i1, int i2) => 
		new ValueList { length = 3, i0 = i0, i1 = i1, i2 = i2 };

	public static ValueList List(int i0, int i1, int i2, int i3) => 
		new ValueList { length = 4, i0 = i0, i1 = i1, i2 = i2, i3 = i3 };
}
```
And the typical use case:
```csharp
class Test
{
	static void Foo(Span<int> span) { }

	static void Main(string[] args)
	{
		Foo(ValueList.List(1, 2, 3));  // safe?
	}
}
```
The codes compile, but I don't know whether or not the temporary struct will be destroyed before calling `Foo`. 
Is there any spec about this?

**Comment by kevinjwz** [comment]
Another unclear case:
```csharp
static void Main(string[] args)
{
	Span<int> span = ValueList.List(1, 2, 3);
	// still safe to use span?
}
```

**Comment by svick** [comment]
I don't believe this is safe. Consider the following code:

```c#
static void Main()
{
    Span<int> span1 = ValueList.List(1, 2, 3);
    Span<int> span2 = ValueList.List(4, 5, 6);
    Console.WriteLine(span1[0]);
}
```

When I run it, it prints `4`, which shows that `span1` is not safe to use, because it got overwritten by another value.

I think this hinges on the lifetime of the temporary variable that is introduced by the compiler when passing a value to an `in` parameter (in this case, the parameter of the conversion operator). But unfortunately, I couldn't find the answer to that question in [the closest thing to the spec for `in` parameters we have at the moment](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.2/readonly-ref).

**Comment by SingleAccretion** [comment]
Yes, the first value list gets overwritten by the second. It is a bit weird though, because initially the lists are actually in different memory locations, but the JIT decides to copy each into the same one. [See on Sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4DCA9lAA4UA21ATgMpcBuFAMbUIAbnRZcBEuSq0AkuS4NmvTgOFj06APQAqdHXwBZMBVIAKAJToAMpgeYQ2ZsQgALbNQAmFO46RnVw8vCD80ewcEZwhiACNQ5gAabEwADwAWDOxdHQhgMEEAa2wwgC9qbAAzBk5sYHcKCGwm7AA2AA5sOIBPYBF0bCGciMcAVmc+Cs4GYmZmLn8HDucoBj5QihTqQTSlzDBnDjAvXxSAbRFmAF1971X1r12U9KRUUcwkaOw02q8wNLbAHDJZIDLOTjUZilYAMCDeYYjSJINoPDY7QGbEY6HTTBjYWbAeruSoAd3cDA4MMKRQAhNoPkgVj8/tQAUC0kiHEhDthjqdMZcIMwANSvDLXLmfaiTX6cZjNNJQKCYFJKlVq5WYJYIb58NZ8bwAR2I2Eu3jS101KqlCFR2ANZp8aTFaSQmCt/05uTx3T6ImqMygiRFCBwsJDGW1aB1AE40dhvKTagihaLxZ70gBmJZZwIOx5JlNO4WupmZt1SrPxgsbIucVNXMuCCsIKUZfOO+uN0vpMMVrPYnQNSpVCicfJ8ppEhiCQRzCgBtOh1VNqO3GMfDI1/XrY2m9Wqkvpt0SpYZKqTA374+urMe604JZjCa1r0XJt9iVSsY1x3L9J1w5XJBBUHpiVHcdJzYad6nxZcMg6JIm06Ddn3uN82UFT80nvSUcQNSpzAjLC+WoUgAHMGgJKoIOqKCiRgycLCzGwPjafN+WdD9e1w78cTYBgwARYTvEhCBmibMY2wsMcJyJagOBocgaLo0DWB8Kd8isOCBVyEcIEqMBIToyEoFIMByGaBhaJHej5NKZhLOwWTiDYNhwPIyEKOnLgfFpNiljaV85SeTEMSCgB2I42U2HibiC3k+DlBUfi1R9HyCmVsF3Q0TRLChPUPREhiWSLvn/EQKDLB8wpKodfV6fpmiqINEl0pswwZSJItfLtkwbY8BzKwQE27W9AIrbIlg6cE33GgC0g6CsxilDpRvmgaexPFsXjSNpsAASGOw7EX0klSh2BhSARJiZznBclyuZDS37bqHFjObcpvQ94o3SJYx3a98sWvDMo+WNL0w54JrdD0ljALME0WoCwpAsC6MM0Cbq0okIwQpCUOWlJ1i4UlOAoShKLouTJ2u6h6U3SIwGi6HsN4vDsAI0mOt2MjKOomy6NIahSWwdTF1u6cEbiGKTiq+K734nRBOEujRPE6zbIu2mFKU8iZ21yoJc0u6OrCBGocq3iEGJrFub4SoGlaDWRGaEjSyjJY4iRt9FoQXa6odp2Lo4AXPA90UsnOkQjJMuyzLMUgtcxq6buaYVLJSckhE8UgGDF8SKG8AMSRMsAiQQ1dPeW3I4gYajM+T7B87F5gGGI3S7IgMAaAdag1k4cCLCkhBAo+OJWcEFh31SNJo0iOJeQAKziSpIkEDI69mdPxZJYpzAo7AwAopPJzssOqM8IW7N13H3swOINsdLCPwtfDhwu8S3KJVo7L4MAbBiBGQUqWaS2dGiCE8K0LeE9sqiUSHtaOHxBD5nbtCKqSxBDfHQaccIG9faQmAFguaghAFsDnm0MMupOgLwcIIGsxEcxoD0DoBkcBByIGwHQbAABvQYwwOEtBUqYcw1gRiIn4WgeqQxuBOVIAAHmIgAPkcpZJA2AAC82AABqgDgG2GnPgQx+QLBIBSAgFIrFxDSJkXIyySjyCqKbm2bReigHUBMcAYx04LAZBSGMFIbQrA2JkRImRcBIpqNIEgc4HpQnDAAL7oGSZuc43BgCcGIIIYAtgwA9EJBYPJBTiDAAANLmG8PgXgJoDYUEAVYDc+Qsk5N0fozx0s0BSMOp3S+DQbE9JUhQVUFBzEUEsRQLMNj0CHSEXgA6VBmAwUEJTAkCxOAVz+PYxRKiLDmDaR4rxOUdKaOUTMw6xgB61B6KYCc7hAH0EhBXag2yLCQlogAVWTmAKotAACCEBCDUCqBYPg+BhlWBSGCvp7gQmHQZLMzhKIDkGNgl4vZQzMAnNUeckWYt3GosnLw/mV8tHYHMS0HA2jhnYESdMtAiKpAHQJR0yc6LO7DJSBypA2KRjHTxSi1lRJiUwrJZYylZLOUtA0dSjRdKEVzORSyo57LMVcqGRSjlY8tE4oZYdAVyrYIivIqS7RWYuVUspVymV0quWuJaG2eVm5GXzMFSq3xHKjzcvVb/cVHLWI6r5fq0WbqjUkuoto/xErqVeptaMu1kq/WDmpYOJ1iSgA).

I guess it is not a bug since we are using `Unsafe`, but this is still rather curious.

**Comment by svick** [comment]
@SingleAccretion I think the JIT is required to do that overwrite, because [the C# compiler decides to reuse the same local variable for both `ValueList`s in IL](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYAYAEMEDoBKArgHbBgC2AprgMID25ADpJQE4DKbAbmAMaUBnANyoM2PEVIVqASVJs6jTqx79hqUXEw1UAb1SZDmMKUwBZAIYmAFAEpMBo/pRHXmdowvEAPCeAA+TAFPYgRMAF5MADULCEJKABkwAWBcJJTrBAAaTDgcgGZbERc3Qw8vX1JA4K8tSJi4xOTU9OBrABYcgFYcgDYix1LBtxgAdiCQhABtdABdYtcAX1RllFQp9mBWQl5gBIsATzpCNv2jk4BpEwATXE4AR3ipWNtZ1BTt3ejY+Na9VAAkH5MBBKMQAObAAAWxSBpjA6CyYGyYDyYHyxUBMHy4l6xiYED4YGAmEUbAswDorHcIUqAWsJm+jVamC49nC/kBALMlHIVIOllYAihsVorEoFMo5WI1nFADNMABVYgCCxy6gAQQE+EocusXFwCNsOQNoIh0KKAI0KAB2NxTN+zUwrQZ8PQ7MCXOIlAA7g6milMLoQWDIVCIphssYsJEEZhFpibXacHiGo7Ay7gQiclmEB6HDaAd6/WmAyTg2awxG8tGI9njGFY2EE9bbTiU/6WZm3Tn4VGs3B817fZ2nRXQ9CI/kczHoznGw2c3VjFoW2sk+2EKmfmXnc1XST67ne4ea1nChFPYXi6PA+PzeHIp1a7HEQ26/3l6iczjYzi14sQA=). Notice that there's only one `[1] valuetype ValueList` and that both use `stloc.1`.

**Comment by kevinjwz** [comment]
I am more concerned about passing `ValueList` as arguments, and a simple test shows it works for now.
```csharp
class Test
{
	static void Foo(Span<int> span1, Span<int> span2)
	{
		Console.WriteLine(span1[0]);  // print 1
		Console.WriteLine(span2[0]);  // print 4
	}

	static void Main(string[] args)
	{
		Foo(ValueList.List(1, 2, 3), ValueList.List(4, 5, 6));
	}
}
```

**Comment by SingleAccretion** [comment]
I think the consensus among people here would be that you shouldn't use this code. This simple and innocent example breaks:
````C#
static void Main(string[] args)
{
   var a = PreserveReference(ValueList.List(1, 2, 3));
   var b = PreserveReference(ValueList.List(4, 5, 6));

   Console.WriteLine(a[0]); //Prints 4
}

static Span<int> PreserveReference(Span<int> span) => span.Slice(0);
````

**Comment by kevinjwz** [comment]
@SingleAccretion That's true, but my motivation is to pass length-variadic tuples as arguments without allocations, and returning references to the arguments is, IMO, pointless in most situations.

**Comment by SingleAccretion** [comment]
I am sorry, that's just me being wrong. The example was indeed OK.

**Comment by kevinjwz** [comment]
In [this documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.2/readonly-ref#temporary-variables), there are 2 sentences about the life time of temporary variables:

>The life time of the argument temporaries matches the closest encompassing scope of the call-site.

>The formal life time of temporary variables is semantically significant in scenarios involving escape analysis of variables returned by reference.

Does the first sentence mean:
```csharp
static Span<int> ToSpan(in ValueList v) => v;

static void Main(string[] args)
{
	if (true)
	{
		var span = ToSpan(ValueList.List(1, 2, 3));
		// ValueList.List(1, 2, 3) still lives
	}
	// ValueList.List(1, 2, 3) is destroyed
}
```
And does it also appliy to implicit conversions?

**Comment by agocke** [comment]
I think the C# language spec states that the `ref` being passed to `CreateSpan` has a global lifetime, so the result is safe to return.

As for whether your code works, the spec would say, `Unsafe` class = lol UB

**Comment by kevinjwz** [comment]
@agocke I think this code didn't introduce a temporary variable.
```cs
public static implicit operator Span<int>(in ValueList v) =>
	MemoryMarshal.CreateSpan(ref Unsafe.AsRef(v.i0), v.length);	
```

**Comment by kevinjwz** [comment]
I think I was working on a wrong direction. 
My goal is to pass variable length tuples as arguments, why not just make the parameter type `in ValueList`?
But it brings a new problem. Suppose `params Span<T>` becomes available, and I want to use it to simplify my `ValueList` struct and lift the limitation on the max length.
```cs
ref struct ValueList
{
	Span<int> span;

	public static implicit operator Span<int>(in ValueList v) => span;

	public static ValueList List(params Span<int> args) => 
		new ValueList { span = args };
}

class Test
{
	static void Foo(in ValueList list) { }

	static void Main(string[] args)
	{
		Foo(ValueList.List(1, 2, 3));  // prevent zero-alloc optimization?
	}
}
```
I am not sure what will happen.
[Current `params Span<T>` proposal](https://github.com/dotnet/csharplang/blob/master/proposals/format.md#extending-params)

**Comment by kevinjwz** [comment]
Though not related to the title, there's a solution looks pretty:
```cs
[StructLayout(LayoutKind.Sequential)]
struct ValueList
{
	int length;
	int i0,i1,i2,i3;

	public static implicit operator Span<int>(in ValueList v) =>
		MemoryMarshal.CreateSpan(ref Unsafe.AsRef(v.i0), v.length);	
	
	// no syntax for one-element tuple
	public static implicit operator ValueList(int value) =>
		new ValueList { length = 1, i0 = value };

	public static implicit operator ValueList(in (int, int) tuple) =>
		new ValueList { length = 2, i0 = tuple.Item1, i1 = tuple.Item2 };

	public static implicit operator ValueList(in (int, int, int) tuple) =>
		new ValueList { length = 3, i0 = tuple.Item1, i1 = tuple.Item2, i2 = tuple.Item3  };

	public static implicit operator ValueList(in (int, int, int, int) tuple) =>
		new ValueList { length = 4, i0 = tuple.Item1, i1 = tuple.Item2, i2 = tuple.Item3, i3 = tuple.Item4 };
}

class Test
{
	static void Foo(in ValueList list1, in ValueList list2) { }

	static void Main(string[] args)
	{
		Foo((1, 2, 3), (4, 5, 6));
	}
}
```

**Comment by kevinjwz** [comment]
However, I still love the idea of using `Span<T>` to receive a tuple of arbitrary length.

**Comment by kevinjwz** [comment]
@agocke 
> I think the C# language spec states that the `ref` being passed to `CreateSpan` has a global lifetime, so the result is safe to return.

Now I understand what you meant. You meant that a `Span<T>` works like a `ref` (managed pointer), so anything the `ref` points to still lives as long as the `Span<T>` lives.
Doesn't that only works when the `ref` points to the interior of a heap object?
```cs
static Span<int> Foo()
{
    int i = 0;
    return MemoryMarshal.CreateSpan(ref i, 1);
}
```
The above compiles, but I don't believe `i` still lives after `Foo` returns.
I think the compiler didn't realize that `MemoryMarshal.CreateSpan` converts a `ref` to a `Span<T>`, just like `Unsafe.AsRef` etc.

**Comment by agocke** [comment]
> Doesn't that only works when the ref points to the interior of a heap object?

Yes, but I believe that's what the ref-return lifetime of `Unsafe.As` is. If a method returns a `ref` and doesn't take any `ref` parameters, what else could it be `ref`-returning? Of course, that's not true for `Unsafe.As` because they violate the safety rules, but that's the point -- they're lying about the safety.

**Comment by kevinjwz** [comment]
@agocke 

> >Doesn't that only works when the ref points to the interior of a heap object?

>Yes, but I believe that's what the ref-return lifetime of Unsafe.As is.

I am not quite understand what you mean.

dotnet/runtime/issues/27295 seems related, but no one talked about temporary struct.

There is a guarantee in C++:

> a temporary bound to a reference parameter in a function call exists until the end of the full expression containing that function call: if the function returns a reference, which outlives the full expression, it becomes a dangling reference.

https://en.cppreference.com/w/cpp/language/reference_initialization

To be clear, that's what I am mainly concerned about.

**Comment by agocke** [comment]
I'm saying that you're currently out of the bounds of the language. The Unsafe and MemoryMarshal APIs violate the Span safety rules, so the language behavior for your program is undefined.

**Comment by kevinjwz** [comment]
@agocke I agree with you. Those API should be used very carefully, in some performance-sensitive code.
But if the lifetime of the struct can be guaranteed, `Unsafe.AsRef<T>(in T)` shouldn't be a problem, since it merely does nothing in IL level.

**Comment by agocke** [comment]
So then the answer to your original question,

> Is there any spec about this?

is yes, but you're out of bounds of the specification. Under the specification there's no way to observe the lifetime of an `in` temporary as a `Span` outside of the method call.

**Comment by kevinjwz** [comment]
@agocke Thanks. 

Perhaps the best choice is to change parameter type to `in ValueList`, since it's clearer than `Span`. Users could unexpectly pass an array to `Span`, making "zero-alloc" meaningless.
